### PR TITLE
reconnect log stream unless container is dead

### LINKED ISSUE #168: Not sending logs to papertrail as of a few days ago
I've been running `latest` in my dev environments for a while now successfully and as of 3/3/2016 I noticed that we weren't getting any logs in papertrail.  The container is started and I when I do a `docker logs logspout` I see only the below (port changed for posting here).  I see nothing in docker.log or anywhere else.  I tried pulling/running the `v2` tag which works for a time and then ultimately crashes when another container is updated and possibly in other scenarios as well (haven't spent time troubleshooting `v2`...would rather get `v3` working).  Using `docker inspect logspout` I didn't see any obvious config differences between the running `v2` and the nonfunctional `v3`.  Any ideas or how I can further troubleshoot?

```
# logspout v3 by gliderlabs
# adapters: syslog raw tls udp tcp
# options : persist:/mnt/routes
# jobs    : pump http[logs,routes]:80
# routes  :
#   ADAPTER	ADDRESS				CONTAINERS	SOURCES	OPTIONS
#   syslog	logs2.papertrailapp.com:12345				map[]
```

### PR DESCRIPTION
**Pull request by bobzoller**
I believe this fixes #168 (at least my own root cause, which seems to be docker/docker#21220).

tested manually; will report back when I've tested it more widely.

### CODE REVIEW

**Comment by johanndt** [comment]
Hi @bobzoller - any more results from your testing? Do you have an image up on docker hub I can also try?

**Comment by johanndt** [comment]
For info, I searched around a bit and found an image at https://hub.docker.com/r/goodeggs/logspout/tags/ with the `reconnect-logs` tag. We've been running it and while some containers seem to be faring better, we are still losing some in papertrail after about half a day.

File: router/pump.go [hunk 1]
```diff
@@ -140,37 +140,58 @@ func (p *LogsPump) pumpLogs(event *docker.APIEvents, backlog bool) {
...
+
+			sinceTime = time.Now()
+
+			container, err := p.client.InspectContainer(id)
+			if err != nil {
+				_, four04 := err.(*docker.NoSuchContainer)
+				if !four04 {
+					assert(err, "pump")
+				}
+			} else if container.State.Running {
```
Comment (rdbaron) [inline router/pump.go]: I found a problem here when testing this fix.  My container restarts so quickly that this check always returns true.  This causes the deletion of the pump below to be skipped and each time my container restarts I get a new one.  This leads to LOTS of duplication.

Comment (bobzoller) [inline router/pump.go]: I haven't run into this problem, but it makes perfect sense.  Off the cuff it seems like we could add a check around line 153 and only create a new pump if one does not exist for that container.  eg `if _, ok := p.pumps[id]; !ok { }`.  Happy to be corrected.

Comment (josegonzalez) [inline router/pump.go]: That should work, mind modifying and testing out that solution?

**Comment by bobzoller** [comment]
I've pushed a new version of docker image `goodeggs/logspout:reconnect-logs` with the fix in 13d1129.  It's running fine on my dev cluster, but perhaps @rdbaron could attempt to reproduce the issue and let us know if it is fixed?

**Comment by rdbaron** [comment]
@bobzoller: will do, I'll let you know once I have.

**Comment by rdbaron** [comment]
@bobzoller; I can confirm that this fixes the issue with duplicating the pump when the container restarts quickly.

**Comment by bobzoller** [comment]
great, thanks!  @josegonzalez seems like this might be ready to merge.

**Comment by josegonzalez** [comment]
Works for me :)

**Comment by josegonzalez** [comment]
Thanks to everyone for being diligent about this PR. Please direct me to any other ones we need to shepherd to greatness :)

**Comment by mattaitchison** [comment]
I am working on a solution for #187 which may also take care of this. 

I also think there are some edge cases this doesn't take care of.

**Comment by josegonzalez** [comment]
@mattaitchison Feel free to override me on this PR, I consider logspout your baby :)

**Comment by bobzoller** [comment]
yes @mattaitchison , and I was fairly certain there were some edge cases this didn't take care of ;)

**Pull request merged**